STN-126 :  Paragraph Postcard

heatherdesigns opened the pull request:
# [STN-126](https://sparkbox.atlassian.net/browse/STN-126) READY FOR REVIEW

## Summary
Add support for the Postcard Paragraph type to the colorful theme. This includes the horizontal and vertical card displays. The postcards will display as simplified versions of the full Card patterns, but should use the same UI Patterns.

Work in this PR includes:
* A new arrow icon mixin
* Moving the headings for all `card` UI Patterns into the templates.
* A refactor of all instances of `URL` to use the existing `Button` field across all `card` UI Patterns.

## Need Review By (Date)
3/10/2020

## Urgency
High

## To test this PR you will need to update your configuration for the Vertical card, Horizontal card, Date Stacked Vertical, and Structured cards.

### Link to Content
Edit view > Format > Show : Settings
Update `Content : Link to Content` Destination to `Button`

### Title
Edit view > Fields > Content : Title
Update `Formatter` to `Plain Text`
* This prevents the title from being wrapped in two h2 tags.
Update `Style Settings`. Unselect `Customize field HTML`. All boxes should be unchecked.

## Steps to Test
1. In the CLI, run `npm run test` and verify all tests pass.
2. Go to http://economics.suhumsci.loc
3. Create a Vertical Postcard
    * Create a basic page
    * In the `Components` block, select `Add Postcard` in the dropdown
        * Create a `Display` with a `Vertical Card`
        * Add an image, title, body copy and link
4. Create a Horizontal Postcard (repeat steps above expect select `horizontal card`)
5. Create a Row of 3 Vertical Postcards
    * In the `Components` block select `Add Row`
    * Then select `Add Postcard`
6. Create a Row of 2 Horizontal Postcards (repeat steps above expect select `horizontal card`)
7. Review pages with the following views: horizontal cards, vertical cards, date stacked vertical cards, and structured cards. Verify the styles look as expected. Also verify the link to content arrow continues to work as expected.

## PR Checklist
- [PR Checklist](https://gist.github.com/sherakama/0ba17601381e3adbe0cad566ad4d80a5)
- [Sparkbox PR Checklist](../docs/SparkboxPRChecklist.md)

Commit:
feat: add support for postcard paragraph type

JIRA STN-126 https://sparkbox.atlassian.net/browse/STN-126

Commit:
feat: create mixin for the arrow icon used for read more links

bryanbraun committed:
WIP: Do the field refactoring discussed in our QA meeting

Commit:
feat: refactor titles for vertical, horizontal, date stacked, and structured cards

pookmish commented:
[Tugboat](https://tugboat.qa) has finished building the preview for this pull request!

Link:
- https://pr513-mcicvrur9kchw6tm86cro4jdggjosijd.tugboat.qa


Dashboard:
- https://dashboard.tugboat.qa/5e65530f8bd53e2907eb0181

Commit:
wip: debugging arrow links on vertical and horizontal cards (and vertical postcards)

Commit:
feat: Continued work on arrow links. Horizontal and Vertical views display as expected. Vertical postcard views are a WIP.

Commit:
feat: fix double arrow icons on the vertical postcards

heatherdesigns commented:
## IE11 Issues: 
* Horizontal Postcards missing images. 
* Horizontal Postcards in row have overlaying content.
<img width="992" alt="IE horizontal cards missing images" src="https://user-images.githubusercontent.com/12678977/76252279-dc225580-621e-11ea-99aa-6403872e67c4.png">

silverli commented:
@heatherdesigns after tomorrow, let's export these new config settings

silverli commented on docroot/themes/humsci/humsci_basic/src/scss/components/_postcard.scss:
```diff
@@ -0,0 +1,107 @@
+    }
+
+    &:hover,
+    &:focus {
```
for all hover effects over the arrow, can we add: `cursor: pointer`?

silverli commented on docroot/themes/humsci/humsci_basic/patterns/horizontal-card/horizontal-card.html.twig:
```diff
@@ -52,8 +54,11 @@
-    {% if url %}
-      <a href="{{ url }}" class="hb-card__decorative-link" aria-label="{{ url }}"></a>
+    {% if button %}
+      {# Do we need this wrapper for styling? #}
```
Yes? I think 😂

silverli commented on docroot/themes/humsci/humsci_basic/patterns/structured-card/structured-card.html.twig:
```diff
@@ -18,7 +18,9 @@
   <div class="hb-card__content">
     {% if title %}
-      <div class="hb-card__title">{{ title }}</div>
+      <div class="hb-card__title">
```
so the structured card doesnt have an arrow right?

heatherdesigns replied:
Correct. The structured card is used for `people` which does not have an arrow.

silverli commented on docroot/themes/humsci/humsci_basic/patterns/vertical-card/vertical-card.html.twig:
```diff
@@ -29,8 +31,11 @@
-    {% if url %}
-      <a href="{{ url }}" class="hb-card__decorative-link" aria-label="{{ url }}"></a>
+    {% if button %}
+      {# Do we need this wrapper for styling? #}
```
yes? check out this comment :)

silverli commented on docroot/themes/humsci/humsci_basic/patterns/vertical-card/vertical-card.ui_patterns.yml:
```diff
@@ -23,17 +23,6 @@ vertical_card:
-      preview: "URL to the item"
-      type: text
-    # The following are used in the original theme but not here
-    author:
```
we can remove author and year?

silverli replied:
(like is H&S ok with that)?

heatherdesigns replied:
I added them back in. It doesn't hurt anything and it may be nice to have for future reference.

silverli commented on docroot/themes/humsci/humsci_basic/src/scss/components/_card.scss:
```diff
@@ -207,23 +207,25 @@
+    }
+
     h2,
     h3,
```
now you can delete the other headings I think!

silverli replied:
(since it just uses h2 now?

heatherdesigns replied:
Oh yeah! 😆

silverli commented on docroot/themes/humsci/humsci_basic/src/scss/components/_card.scss:
```diff
@@ -249,14 +251,34 @@
+      width: 100%;
+      margin-left: hb-calculate-rems(-44px);
+      text-decoration: none;
+      z-index: $hb-z-index-decorative-link;
```
hm why does it need zindex? can you leave a comment?

heatherdesigns replied:
I'm glad you made this request. I found that clicking on the arrow icon in the horizontal card wasn't working and that I needed to add this to another class as well. :tada:

silverli commented on docroot/themes/humsci/humsci_basic/src/scss/components/_card.scss:
```diff
@@ -249,14 +251,34 @@
 
       @include hb-colorful {
-        background: no-repeat left / 55% svg('<svg viewBox="0 0 23 24"><path fill="#{hb-colorful-variation(secondary)}" d="M11 23l-2-2 7-7H0v-3h16L9 3l2-2 10 10 1 1-11 11z"/></svg>');
+        @include hb-link-arrow-icon($color: hb-colorful-variation(secondary));
```
nice!

silverli commented on docroot/themes/humsci/humsci_basic/src/scss/components/_postcard.scss:
```diff
@@ -0,0 +1,107 @@
+    }
+
+    &__subcontent {
+      order: 2; // Note: From an accessibility standup, this not recommended. It takes the content out of it's normal flow in the DOM.
```
standpoint, not standup I think :)

heatherdesigns replied:
Oh my goodness!!!! 🤣

silverli commented on docroot/themes/humsci/humsci_basic/src/scss/components/_postcard.scss:
```diff
@@ -0,0 +1,107 @@
+.ptype-hs-postcard {
+  .hb-card {
```
can you add a little comment here about how these are all override styles to get the postcard paragraph type working?

silverli commented on docroot/themes/humsci/humsci_basic/src/scss/components/_postcard.scss:
```diff
@@ -0,0 +1,107 @@
+.field-hs-postcard {
+  &-title {
+    h2,
+    h3,
```
do you need all levels of headings now?

silverli replied:
since it's an h2 now?

silverli commented on docroot/themes/humsci/humsci_basic/src/scss/components/_postcard.scss:
```diff
@@ -0,0 +1,107 @@
+
+    &.decanter-button {
+      @include unbuttonize;
+      // margin-left: hb-calculate-rems(-44px);
```
do you want this?

heatherdesigns replied:
Nope! :smile:

silverli commented on docroot/themes/humsci/humsci_basic/src/scss/tools/_mixins.icons.scss:
```diff
@@ -0,0 +1,30 @@
+}
+
+@mixin hb-link-arrow-icon($color: $hb-color--black) {
+  background: no-repeat left / 55% svg('<svg viewBox="0 0 23 24"><path fill="#{$color}" d="M11 23l-2-2 7-7H0v-3h16L9 3l2-2 10 10 1 1-11 11z"/></svg>');
```
Yay! is there anything else that can be extracted here? or it's probably better to keep all the transforms on hover in another area?

heatherdesigns replied:
I almost went the route of creating a mixin that included the transform, hover, and focus states, but then I looked at how this sits inside of the `hb-colorful` theme and decided to keep the scope narrowed down to just the icon.

![Screen Shot 2020-03-10 at 10 35 28 AM](https://user-images.githubusercontent.com/12678977/76323361-ff024780-62ba-11ea-9c24-c79317c7953b.png)

Perhaps when we implement additional themes there will be an opportunity to refactor.

silverli replied:
ok sounds good!

silverli reviewed (changes requested):
Looking really great. A couple small code asks and questions. Additionally I had the following notes:

Date stacked vertical card is not outputting arrow correctly for me - maybe I am doing something wrong but the link text is showing. (no longer an issue! I had a checkbox for plan text checked)

👍 Also, if the postcard is nestled with other content, like a textarea, we might want to add margin-top:
(looks great now)

<img width="1008" alt="Screen Shot 2020-03-09 at 8 20 44 PM" src="https://user-images.githubusercontent.com/12848168/76270414-9334c600-624b-11ea-87ae-d9d3a91199bf.png">


For browser testing I did the following:

- Safari ✅ 
- Firefox ✅ 
- Chrome ✅ 
- IE 11 (I see the same issues with horizontal card-- weird!) but vertical card looks amazing!

Commit:
feat: implement review feedback

silverli commented on docroot/themes/humsci/humsci_basic/src/scss/components/_card.scss:
```diff
@@ -59,6 +59,7 @@
         color: $hb-color--black;
         border-bottom: 1px solid $hb-color--black;
         box-shadow: none;
+        cursor: pointer;
```
wait a tags shouldnt need this though?

heatherdesigns replied:
I got over zealous! 😆

silverli commented on docroot/themes/humsci/humsci_basic/src/scss/components/_card.scss:
```diff
@@ -262,12 +260,13 @@
       &:focus {
         box-shadow: unset;
         color: transparent;
+        cursor: pointer;
```
same conundrum here

silverli replied:
I just wanted to make sure the arrow had it

silverli commented on docroot/themes/humsci/humsci_basic/src/scss/utilities/_wysiwyg-text-area.scss:
```diff
@@ -1,6 +1,6 @@
 // WYSIWYG text area styles
 .field-hs-text-area {
-  margin: 0 hb-calculate-rems(48px);
+  margin-bottom: hb-calculate-rems(48px);
```
🎉

heatherdesigns commented:
TODO: also document in the CSS that the post card places the arrow in the subcontent for horizontal card

Commit:
feat: squash IE bug

- horizontal card images in IE need a flex-shrink property to prevent them from collapsing.
- prevent the max-width property on content from rendering in IE. This allows for flexbox fallback properties to work correctly.

Commit:
feat: add arrow icon to the icon variables

- update arrrow icon mixin and all calls to it to have a similar naming convention as the arrow icon varible

silverli commented on docroot/themes/humsci/humsci_basic/src/scss/components/_postcard.scss:
```diff
@@ -1,6 +1,10 @@
+  // Override the hb-card styles so that they display correctly within paragraph postcard components.
   .hb-card {
+    &__img {
+      flex-shrink: 0; // prevent horizontal card image from collapsing
```
oh interesting 🤔 very cool :)

silverli commented on docroot/themes/humsci/humsci_basic/src/scss/components/_postcard.scss:
```diff
@@ -10,6 +14,7 @@
       margin-bottom: 0;
     }
 
+    // For horizontal postcards, Content to Link is mapped to the Time field, this places the Content to Link within the subcontent.
```
or maybe instead of content to link, it would be clearer to say "arrow"?

silverli commented on docroot/themes/humsci/humsci_basic/src/scss/components/_card.scss:
```diff
@@ -201,29 +201,30 @@
-        max-width: 100%;
+
+        @supports (max-width: 100%) {
+          max-width: 100%; // Apply max-width to all browsers except IE. This allows the fallback settings written in flex-column to work correctly.
```
so weird! why does this break ie?

heatherdesigns replied:
I found [this article](https://alastairc.ac/2014/08/flexbox-ie11-bugs/) where they were experiencing the same bug. Since we added max-width to overwrite the styles written for `flex-column` I decided to remove it and see what happened. I was pleasantly surprised to find that things look better in IE without it.

silverli commented on docroot/themes/humsci/humsci_basic/src/scss/components/_card.scss:
```diff
@@ -265,11 +287,13 @@
 
     &:hover,
     &:focus {
+      cursor: pointer;
```
👍

silverli commented on docroot/themes/humsci/humsci_basic/src/scss/components/_postcard.scss:
```diff
@@ -0,0 +1,111 @@
+      &:focus {
+        box-shadow: none;
+        color: transparent;
+        cursor: pointer;
```
hmm I dont know that the cursor is needed here!

silverli commented on docroot/themes/humsci/humsci_basic/src/scss/components/_postcard.scss:
```diff
@@ -0,0 +1,111 @@
+
+    &:hover,
+    &:focus {
+      cursor: pointer;
```
can you verify this cursor: pointer is ok / needed here?

silverli commented on docroot/themes/humsci/humsci_basic/src/scss/tools/_mixins.icons.scss:
```diff
@@ -0,0 +1,30 @@
+}
+
+@mixin hb-icon-link-arrow($color: $hb-color--black) {
+  background: no-repeat left / 55% svg('<svg viewBox="0 0 23 24"><path fill="#{$color}" d="M11 23l-2-2 7-7H0v-3h16L9 3l2-2 10 10 1 1-11 11z"/></svg>');
```
wait! shouldnt this use your new variable? `$hb-icon-link-arrow`

silverli replied:
oh i see.. you're passing dynamic colors in

silverli replied:
nevermind!

silverli reviewed (changes requested):
This is so close!! Just a couple last small things. The postcard looks stunning in IE now :) 🎉

Commit:
feat: minor tweaks

- clarify comments regarding use of decorative links / arrows
- remove unnecessary properties

silverli commented on docroot/themes/humsci/humsci_basic/src/scss/components/_postcard.scss:
```diff
@@ -14,7 +14,9 @@
     }
 
-    // For horizontal postcards, Content to Link is mapped to the Time field, this places the Content to Link within the subcontent.
+    // For horizontal postcards, Content to Link (which is represented by
```
awesome

silverli reviewed (changes requested):
gah! one last annoying thing but the hcards still dont have cursor: pointer hover on the arrows :(

bryanbraun committed:
fix: restore the pointer icon when hovering over the svg

silverli reviewed (approved):
Looks good! thanks for fixing that hover state
